### DISCUSSION TITLE
Script Categories

### DISCUSSION THREAD

**Discussion by Holzchopf** [flp-community-scripts/community-scripts#7] [Ideas]
There are a lot of piano roll scripts and ideally they're presented to the end user in a way so they can easily find what they're looking for. One approach for this is to categorize the scripts by their main use and reflect this categorization in the folder structure, which in turn determines the menu structure in the Piano roll > Tools > Scripts section.

Some thoughts:
- our collection should not interfere with built-in scripts
- our collection should not interfere with scripts from the FL Studio content library
- there should not be too much "menu diving"
- the categorization should be sensible and convenient
- overall, we'd like to offer a great UX

The first two points could be addressed by using a dedicated folder for the collection in the top-level folder. This could be `Community scripts` for example. But right now, that name does not matter, since the collection is offered simply as a git repository and users who'd like to install it can (or have to) manually name the target folder.

The next point concerning "menu diving" could be addressed by flattening out categories; instead of having categories with sub-categories, the categories could be put in the same level and distinguished with an affix. For example: instead of `Generate > Chords > scripts...` and `Generate > Melody > scripts...`, simply `Generate (chords) > scripts...` and `Generate (melody) > scripts...`.

The next point concerning convenience is by far the most complicated to resolve. It's a plain UX topic and as we all know, code writers aren't necessarily the best UX designers [citation needed]. And even if UX designers were involved, they'd probably disagree on many points. The best is to try to copy similar software solutions as best as possible, because similarity helps the user to find its way around. Even though there's no other "flpianoroll scripts menu" out there, basic categorization for different editing tasks *can* be found in some software. But, there's an additional difficulty: The menu entries are simply listed alphabetically by FL Studio.

My first approach was to categorize into "generates score", "edits score" (like, adds and removes notes) and "modifies note properties". We then quickly found `Adjust`, `Edit`, `Generate (chords)`, `Generate (melody)`, `Generate (rhythm)`, `Markers`, `Select` and `Utils`, which I personally like, since its alphabetical order also kinda groups the scripts/categories into "do a little bit something", "do something", "DO SOMETHING!!!" followed by utilities for markers, selection and whatnot.

In contrast, aiming for a great UX is quite simple (in theory): Just forget for a second that you're developer and ask yourself "what would I expect if I was the user?". This affects us in two ways:
- What will the user be looking if they want to do a certain thing?
- Where will they be looking for that?

Answering this well will shorten the time a user is looking for something and in turn make it less frustrating to work with the collection. Concretely: Script names and category names should be self explanatory. Yes, this goes hand in hand with the categorization topic. But maybe just re-read the paragraphs about categorization again with "What would the user want?" in mind...

**But enough of my talk. Let's discuss some things!**

**Comment by Holzchopf** [comment]
I'll start:

When sorting my scripts, I ended up moving quite a lot of them into `Adjust` and `Edit`, 13 and 16 of them, respectively. Some of them do really basic everyday stuff, like fade in or fade out, others are for very rare occasions only, like swapping the mod x and mod y note properties. Some adjust one property exclusively, others adjust things like note number and length as well, where it's discussable whether these only adjust notes or in fact alter the score altoghether. Some run without user input, others are really sophisticated and need intricate setup.

With edit scripts it's no easier: some add notes, some move notes up/down, some move notes left/right, some move notes randomly, some cut up existing notes, some do all of the above.

I think, the two categories `Adjust` and `Edit` could use finer granulization as well, just like the generative scripts.

For the `Edit` category, I could imagine categories like
- `Edit (effects)`
- `Edit (slice)`
- `Edit (transform)`

But I'm not sure whether the "effects" suffix is clear enough?

The `Adjust` category could maybe be split into the property they adjust (velocity, panning, color, tuning...), plus one for scripts that touch multiple of them?

**Comment by fruitybagel** [comment]
Hmmm yess yess, I noticed a lot of `Select` scripts so I did those but the suggestions for `Markers`, `Edit (effects)`, `Edit (slice)` and `Edit (transform)` is interesting. However I do think we should be more obvious with these. Cuz even I'm not entirely sure about the complete distinction between `Adjust` and `Edit` in some cases

**Comment by fruitybagel** [comment]
I also created a `Color` category I guess Xd Idk maybe it's okay to have like a handful of categories, they shortcut having to understand or dig in for a doc

**Comment by fruitybagel** [comment]
Also yeah I think if we are changing the categories ideally we update the Categories.md under repo-spec for book keeping

**Comment by Holzchopf** [comment]
My initial thought was that we could use "scripts that add, move or remove notes" could be distinct from "scripts that only modify existing notes properties". But I agree, the distinction into "Adjust" and "Edit" isn't very obvious.

**Reply by fruitybagel** [reply]
Ahhh yeah,  so if the script moves the notes across pitch I guess then it's `Edit` I assume? I think I might have some misplaced archive scripts here and there
